### DISCUSSION TITLE
No serial comms to RP6502

### DISCUSSION THREAD

**Discussion by mickstone** [picocomputer/community#105] [Q&A] [answered]
Hi All,

Another day and another issue. I've had some fun over the last few days getting my build environment working and can now compile binaries for the RP6502. I can't send the program to the PicoComputer via serial though.

I can confirm I can connect to the PicoComputer via miniterm or screen, but the rp6502.py script fails to communicate.

The port looks okay in ,rp6502 and is set as /dev/ttyACM0 which appears to work fine connecting via minicom. I did have to adjust permissions to allow access.

I'm not certain if my issue is a result of not being able to install pyserial. I can confirm that python3-serial is installed and "appears" to work, but not smart enough to be completely sure.

I can see initially there is an issue with an invalid escape character that I can get to go away if I swap ^\$ to ^/$ but still no comms with the RP6502.

The following is the terminal output building "advent"

Executing task: python3 rp6502.py -c /home/mick/Build/adventure/.rp6502 run /home/mick/Build/adventure/build/advent.rp6502 

/home/mick/Build/adventure/tools/rp6502.py:180: SyntaxWarning: invalid escape sequence '\$'
  str = re.sub('^\$', '0x', str)
/home/mick/Build/adventure/tools/rp6502.py:277: SyntaxWarning: invalid escape sequence '\$'
  str = re.sub('^\$', '0x', str)
[rp6502.py] Loading ROM /home/mick/Build/adventure/build/advent.rp6502
[rp6502.py] Opening device /dev/ttyACM0
Traceback (most recent call last):
  File "/home/mick/Build/adventure/tools/rp6502.py", line 344, in <module>
    exec_args()
  File "/home/mick/Build/adventure/tools/rp6502.py", line 294, in exec_args
    mon.send_break()
  File "/home/mick/Build/adventure/tools/rp6502.py", line 49, in send_break
    self.send_break(duration, retries-1)
  File "/home/mick/Build/adventure/tools/rp6502.py", line 48, in send_break
    raise te
  File "/home/mick/Build/adventure/tools/rp6502.py", line 44, in send_break
    self.wait_for_prompt(']')
  File "/home/mick/Build/adventure/tools/rp6502.py", line 110, in wait_for_prompt
    raise TimeoutError()
TimeoutError

 *  The terminal process "/usr/bin/bash '-c', 'python3 rp6502.py -c /home/mick/Build/adventure/.rp6502 run /home/mick/Build/adventure/build/advent.rp6502'" terminated with exit code: 1.

**Comment by netzerohero** [comment]
Be sure to exit minicom fully before using the python script. If minicom is up and has been allocated the serial-port, the python script cannot allocate the serial-port for its usage.

**Reply by mickstone** [reply]
Thanks, noted and understood

**Comment by vruumllc** [comment]
Can you run or copy the resulting binary manually via the command line?

```
tonyvr@tonyvr-laptop:~$ cd projects/RP6502/vscode-cc65/
tonyvr@tonyvr-laptop:~/projects/RP6502/vscode-cc65$ python tools/rp6502.py run build/hello.rp6502
[rp6502.py] Loading ROM build/hello.rp6502
[rp6502.py] Opening device /dev/ttyACM0
tonyvr@tonyvr-laptop:~/projects/RP6502/vscode-cc65$ python tools/rp6502.py upload build/hello.rp6502
[rp6502.py] Opening device /dev/ttyACM0
[rp6502.py] Uploading build/hello.rp6502
```

**Reply by vruumllc** [reply]
It is interesting that you are getting the escape sequence warning. That was an old bug that was fixed in the tools/ script. How old is your vscode-cc65 project? Make sure it matches the latest in the repository.

**Reply by mickstone** [reply]
I've just confirmed it is the latest release. Have the issue in the hello world project and the adventure project.

I've tried running from a command line with same result.

mick@mick-XPS-9315:~/Build/PicoComputer$ python3 tools/rp6502.py run build/hello.rp6502
[rp6502.py] Loading ROM build/hello.rp6502
[rp6502.py] Opening device /dev/ttyACM0
Traceback (most recent call last):
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 400, in <module>
    exec_args()
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 342, in exec_args
    mon.send_break()
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 49, in send_break
    self.send_break(duration, retries - 1)
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 48, in send_break
    raise te
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 44, in send_break
    self.wait_for_prompt("]")
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 110, in wait_for_prompt
    raise TimeoutError()
TimeoutError

Can one of you confirm that a fresh copy of advent or hello compiles and uploads? i.e. the rp6502.py script should be the same as ones I'm running.

**Reply by vruumllc** [reply]
Let's just forget about the rp6502.py script for a second and work with python interactively.

1. First, why don't you restart Ubuntu, and then open a command prompt (and NOT minicom, or whatever you are using).

2. Unplug the usb cable from the RP6502, then in the command prompt, type:

```
tonyvr@tonyvr-laptop:~$ ls /dev/ttyA*
ls: cannot access '/dev/ttyA*': No such file or directory
```

3. Now plug the usb cable back into the VGA pico on the RP6502, and type:

```
tonyvr@tonyvr-laptop:~$ ls /dev/ttyA*
/dev/ttyACM0
```

4. If all is good so far, let's start python, and manually send the "help" command to the RP6502 via the serial port:

```
tonyvr@tonyvr-laptop:~$ python
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import serial
>>> s = serial.Serial()
>>> s.setPort("/dev/ttyACM0")
>>> s.timeout = 0.5
>>> s.baudrate = 115200
>>> s.open()
>>> s.write(b"help\r")
5
>>> s.close()
>>> quit()
tonyvr@tonyvr-laptop:~$ 
```

5. Did this work? If not, try running python as sudo, then type the same commands

```
tonyvr@tonyvr-laptop:~$ sudo python
[sudo] password for tonyvr: 
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import serial
>>> s = serial.Serial()
>>> s.setPort("/dev/ttyACM0")
>>> s.timeout = 0.5
>>> s.baudrate = 115200
>>> s.open()
>>> s.write(b"help\r")
5
>>> s.close()
>>> quit()
tonyvr@tonyvr-laptop:~$ 
```

6. Did that work?

**Reply by vruumllc** [reply]
One note about my commands above: I have my "python" command automatically aliased to "python3". I see you are typing python3 directly, and that is fine.

**Comment by mickstone** [comment]
So that works okay, I had the PicoComputer display help on the screen. I didn't run as sudo.

<pre><font color="#26A269"><b>mick@mick-XPS-9315</b></font>:<font color="#12488B"><b>~</b></font>$ ls /dev/ttyA*
ls: cannot access &apos;/dev/ttyA*&apos;: No such file or directory
<font color="#26A269"><b>mick@mick-XPS-9315</b></font>:<font color="#12488B"><b>~</b></font>$ ls /dev/ttyA*
<span style="background-color:#171421"><font color="#A2734C"><b>/dev/ttyACM0</b></font></span>
<font color="#26A269"><b>mick@mick-XPS-9315</b></font>:<font color="#12488B"><b>~</b></font>$ python
Command &apos;python&apos; not found, did you mean:
  command &apos;python3&apos; from deb python3
  command &apos;python&apos; from deb python-is-python3
<font color="#26A269"><b>mick@mick-XPS-9315</b></font>:<font color="#12488B"><b>~</b></font>$ python3
Python 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] on linux
Type &quot;help&quot;, &quot;copyright&quot;, &quot;credits&quot; or &quot;license&quot; for more information.
&gt;&gt;&gt; import serial
&gt;&gt;&gt; s = serial.Serial()
&gt;&gt;&gt; s.setPort(&quot;/dev/ttyACM0&quot;)
&gt;&gt;&gt; s.timeout = 0.5
&gt;&gt;&gt; s.baudrate = 115200
&gt;&gt;&gt; s.open()
&gt;&gt;&gt; s.write(b&quot;help\r&quot;)
5
&gt;&gt;&gt; s.close()
&gt;&gt;&gt; quit
Use quit() or Ctrl-D (i.e. EOF) to exit
&gt;&gt;&gt; quit()
</pre>

**Reply by vruumllc** [reply]
OK, that is really excellent news, as that is exactly what the tools/rp6502.py script does.

It seems like some process was keeping the serial port from opening earlier.  (maybe a hung minicom session?)

Anyway, cd to the vscode-cc65 project directory, and try running the hello binary again:

```
tonyvr@tonyvr-laptop:~/projects/RP6502/vscode-cc65$ python3 tools/rp6502.py run build/hello.rp6502
[rp6502.py] Loading ROM build/hello.rp6502
[rp6502.py] Opening device /dev/ttyACM0
tonyvr@tonyvr-laptop:~/projects/RP6502/vscode-cc65$ 
```

**Reply by mickstone** [reply]
I wish that worked, but it didn't and if you see below, the error is the same. Right after running that I tried your direct approach with python and was able to send "help" via /tty/ACM0 to the PicoComputer,

Also to confirm that the port was not in use before running the below I popped the cable and reinserted to make sure the port was free.



mick@mick-XPS-9315:~/Build/PicoComputer$ python3 tools/rp6502.py run build/hello.rp6502
[rp6502.py] Loading ROM build/hello.rp6502
[rp6502.py] Opening device /dev/ttyACM0
Traceback (most recent call last):
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 400, in <module>
    exec_args()
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 342, in exec_args
    mon.send_break()
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 49, in send_break
    self.send_break(duration, retries - 1)
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 48, in send_break
    raise te
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 44, in send_break
    self.wait_for_prompt("]")
  File "/home/mick/Build/PicoComputer/tools/rp6502.py", line 110, in wait_for_prompt
    raise TimeoutError()
TimeoutError
mick@mick-XPS-9315:~/Build/PicoComputer$ ^C




mick@mick-XPS-9315:~/Build/PicoComputer$ python3
Python 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import serial
>>> s = serial.Serial()
>>> s.setPort("/dev/ttyACM0")
>>> s.timeout = 0.5
>>> s.baudrate = 115200
>>> s.open()
>>> s.write(b"help\r")
5

**Reply by vruumllc** [reply]
Wow, that is really puzzling! I did verify that I, too, am using the latest rp6502.py script from the vscode-cc65 project, so ours should be the same.

I really don't know what else to suggest. Maybe try increasing the DEFAULT_TIMEOUT value in the rp6502.py script?

**Reply by mickstone** [reply]
Yeah, I tried that too. Changed it to 5 seconds and it took longer to fail lol.

I've learned more in the last week of poking around than I have over the last 20 years regarding Ubuntu and coding. Never touched python before now. I did dabble a little with machine code programming on the C64 and with C a little later with the Amiga, but it has been a long break and most of my brain cells have died since then. haha.  Thanks for trying with me :)

**Reply by vruumllc** [reply]
Actually, going back over your error messages, it looks like the open() was working fine, and the timeout was waiting for the "]" prompt. My bad, sorry. I still don't have any more ideas, though.

What RP6502 firmware version are you running? https://github.com/picocomputer/rp6502/releases

**Reply by vruumllc** [reply]
Oh, well, at least you can compile, then transfer the binary manually via the usb stick. I wish I could come up with something better.

Don't forget to check out my stuff at https://github.com/vruumllc

**Reply by mickstone** [reply]
Running 0.7 firmware in the picos

Many thanks for giving me some of your time. I really appreciate it.

**Reply by vruumllc** [reply]
Sorry, one last suggestion. Did you try running the python command as sudo? Maybe there is some weird Linux permission thing going on where writes work and reads don't (desperation time here...).

`mick@mick-XPS-9315:/Build/PicoComputer$ sudo python3 tools/rp6502.py run build/hello.rp6502`

**Reply by mickstone** [reply]
Unfortunately, same result :(

**Comment by netzerohero** [comment]
@mickstone 

I've read through and followed @vruumllc's and your attempts to debug your serial issue.  Tony's suggestions have been quite thorough.

IIRC, I had (what I believe to be) the same issue a few months ago (when I was beating on the HW and using the pico6502's tool-chain more rigorously than recently).  I racked my brains out when the issue affected me, that only the night before I knew had worked fine.  I think I then recalled that my Linux OS kernel had updated overnight; it was a "minor' update.  I manually reverted the kernel upgrade to 1-previous; my problem resolved (localized in my case to kernel / python3 issue.)  (I run LMDE Linux Mint Debian Edition, not Ubuntu.)  As a suggestion, perhaps try rolling -back to another kernel (or move forward) with your host OS kernel.  (I know it sounds crazy, but I believe the regression I experienced then I isolated to the kernel update.)


Update:  After I wrote the above, I tracked-down my log of my recollection;  the issue was with running kernel: linux-image-6.1.0-18-amd64.
At the time (Feb-2024) reverting to linux-image-6.1.0-17-amd64 resolved the issue.  I confirmed the issue on two separate laptops with the problematic linux-image-6.1.0-18-amd64.  It didn't play nice with the python3 serial-tools (for reasons unknown).

Attached are my log notes of that day. 
[python3_upload_troubles (netzerohero).txt](https://github.com/picocomputer/community/files/15156114/python3_upload_troubles.netzerohero.txt)

Hope this helps.

**Reply by mickstone** [reply]
Thanks for this. I will likely need to start from scratch again. I'll look at my kernel version when I can get away from work.
I'm running Ubuntu 24.04 build freshly installed.
I could go back to earlier LTS versions if you are able to make a recommendation?

**Reply by netzerohero** [reply]
Update: Being curious, today (2-May-2024) I tried to load the hello-world using the python serial-tools.  Turns-out: The python load script fails; I have the same issue when using the latest LDME-6 kernel: 6.1.0-20-amd64 and with 6.1.0-18-amd64.  

However with kernel  6.1.0-17-amd64 the load script works.

Background and details:
I use the Debian-Edition of Linux Mint (LMDE 6), which uses an older kernel (6.1.0-x) than Ubuntu 24.02 LTS.  (Use: "uname -r" in a command terminal.)  A web search returns that both the Ubuntu 24.04 and 22.04 LTSes use newer kernels (6.8 and 6.5 respectively) than LMDE-6's 6.1.0-x.  I have the same difficulties you are experiencing with the python3 serial-tool when using kernels 6.1.0-18 and 6.2.0-20.  My (Feb-2024) kernel 6.1.0-17 works with the pico6502 tool-chain.

I first downloaded the latest hello-world package with the tools; then ran three tests.

Attached is my test-log file.
[rp6502tool_issue.txt](https://github.com/picocomputer/community/files/15192415/rp6502tool_issue.txt)

Here is the last test (failure):

Test-3:
$ uname -r
6.1.0-20-amd64
netzerohero@E6400:~/Documents/computer/rp6502computer/vscode-cc65-main/tools
$ python3 rp6502.py run ../build/hello.rp6502 
[rp6502.py] Loading ROM ../build/hello.rp6502
[rp6502.py] Opening device /dev/ttyACM0
Traceback (most recent call last):
  File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 400, in <module>
    exec_args()
  File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 342, in exec_args
    mon.send_break()
  File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 49, in send_break
    self.send_break(duration, retries - 1)
  File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 48, in send_break
    raise te
  File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 44, in send_break
    self.wait_for_prompt("]")
  File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 110, in wait_for_prompt
    raise TimeoutError()
TimeoutError

------------

**Reply by vruumllc** [reply]
Hi, I use this:
```
tonyvr@tonyvr-laptop:~$ uname -a
Linux tonyvr-laptop 6.5.0-28-generic #29~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Apr  4 14:39:20 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

**Reply by mickstone** [reply]
Thanks Tony, I'll try that.

**Reply by mickstone** [reply]
> Update: Being curious, today (2-May-2024) I tried to load the hello-world using the python serial-tools.  Turns-out: The python load script fails; I have the same issue when using the latest LDME-6 kernel: 6.1.0-20-amd64 and with 6.1.0-18-amd64.  
> 
> 
> 
> However with kernel  6.1.0-17-amd64 the load script works.
> 
> 
> 
> Background and details:
> 
> I use the Debian-Edition of Linux Mint (LMDE 6), which uses an older kernel (6.1.0-x) than Ubuntu 24.02 LTS.  (Use: "uname -r" in a command terminal.)  A web search returns that both the Ubuntu 24.04 and 22.04 LTSes use newer kernels (6.8 and 6.5 respectively) than LMDE-6's 6.1.0-x.  I have the same difficulties you are experiencing with the python3 serial-tool when using kernels 6.1.0-18 and 6.2.0-20.  My (Feb-2024) kernel 6.1.0-17 works with the pico6502 tool-chain.
> 
> 
> 
> I first downloaded the latest hello-world package with the tools; then ran three tests.
> 
> 
> 
> Attached is my test-log file.
> 
> [rp6502tool_issue.txt](https://github.com/picocomputer/community/files/15192415/rp6502tool_issue.txt)
> 
> 
> 
> Here is the last test (failure):
> 
> 
> 
> Test-3:
> 
> $ uname -r
> 
> 6.1.0-20-amd64
> 
> netzerohero@E6400:~/Documents/computer/rp6502computer/vscode-cc65-main/tools
> 
> $ python3 rp6502.py run ../build/hello.rp6502 
> 
> [rp6502.py] Loading ROM ../build/hello.rp6502
> 
> [rp6502.py] Opening device /dev/ttyACM0
> 
> Traceback (most recent call last):
> 
>   File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 400, in <module>
> 
>     exec_args()
> 
>   File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 342, in exec_args
> 
>     mon.send_break()
> 
>   File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 49, in send_break
> 
>     self.send_break(duration, retries - 1)
> 
>   File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 48, in send_break
> 
>     raise te
> 
>   File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 44, in send_break
> 
>     self.wait_for_prompt("]")
> 
>   File "/home/netzerohero/Documents/computer/rp6502computer/vscode-cc65-main/tools/rp6502.py", line 110, in wait_for_prompt
> 
>     raise TimeoutError()
> 
> TimeoutError
> 
> 
> 
> ------------
> 
> 
> 
> 
> 
> 

Thanks. This is exactly what I'm seeing. I'll rebuild with 22.04 over the weekend and see how I go.

**Comment by mickstone** [comment, marked as answer]
Success, many thanks guys for pointing the way.

I've done a clean build of Ubuntu 22.04.4 and managed to get everything working.

Hopefully someone finds this information useful in future builds.

Thanks @vruumllc and @netzerohero appreciate you sticking with me to get this far

**Reply by netzerohero** [reply]
Glad it works for you!

Both you and Tonyvr are running Ubuntu 22.04 and the script works there.  Looks like the root-cause IMHO is still not resolved since the serial-script tool didn't work for you on Ubuntu 24.04.   And the script breaks for me on LMDE 6 when I move beyond kernel 6.1.0-17-amd64.  So I suspect something is still lurking.... (like a race condition on the "]" prompt):

File ".../tools/rp6502.py", line 44, in send_break

self.wait_for_prompt("]")

**Comment by patrick-flynn** [comment]
Hi everybody. I ran into the same issue after I upgraded Ubuntu on my Linux dev system to 24.04, and investigated.

TL;DR: need to use a newer TinyUSB stack (dated after April 30)  in the firmware (VGA at least, both RIA and VGA to be safe?)

More detail:
1. A December 2023 Linux kernel update ([68aad7d](https://github.com/torvalds/linux/commit/66aad7d8d3ec5a3a8ec2023841bcec2ded5f65c9), apparently) "corrects" prior poor behavior of the Linux DCD-ACM driver to make the kernel respect the connected device's capability report for inbound hardware break handling. If the device says "I can't handle breaks", the serial driver on the Linux side won't send them now.
2. This was noticed by TinyUSB users and addressed in ([PR2616](https://github.com/hathach/tinyusb/pull/2616)). The exact fix is to update usbd.h to report the correct capability flags:  About line 220, the `2` at the end of the  `CDC_FUNC_DESC_ABSTRACT_CONTROL_MANAGEMENT` configuration is changed to `6`.
3. I implemented the single-character change in usbd.h and the build script correctly loaded binaries to the R6502 after the updated firmware was installed.

@rumbledethumps is there a chance the firmware can be updated to reference a newer version of tinyusb than 86c416d ?

**Reply by vruumllc** [reply]
A recent update to Ubuntu 22.04.4 LTS also broke my previously working downloads after builds.
Changing the 2 to a 6 in TinyUSB's usbd.h file, as reported above, fixed the problem for me too.

Thanks so much for figuring this out @patrick-flynn!

**Reply by patrick-flynn** [reply]
I'm glad it worked for you as well.  And I'll return the favor - thank you for the bitmap_graphics library. I'm making a lot of use of it in some demos and examples. -- pf

**Reply by WojciechGw** [reply]
@patrick-flynn thank You very much -  it works for me ... after many hours of searching for a solution ;)
```
$lsb_release -a
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal
```
as WSL2 under Windows 10
Linux APACHE 5.15.153.1-microsoft-standard-WSL2

**Reply by rumbledethumps** [reply]
> TL;DR: need to use a newer TinyUSB stack

Version 0.8 has this fix.
